# Data Modeling

---

# 1. Các bước thực hiện

## 1.1 Chọn mô hình phù hợp:

Học có giám sát (Supervised learning):

- Hồi quy (Regression): Sử dụng cho các bài toán dự đoán giá trị liên tục. Ví dụ: linear regression, logistic regression.

- Phân loại (Classification): Sử dụng cho các bài toán phân loại dữ liệu vào các nhóm cụ thể. Ví dụ: decision tree, random forest, SVM, neural network.

Học không giám sát (Unsupervised learning):

- Phân cụm (Clustering): Sử dụng để nhóm các điểm dữ liệu thành các cụm dựa trên sự tương đồng. Ví dụ: K-means, DBSCAN, Hierarchical clustering.

- Giảm chiều (Dimensionality Reduction): Sử dụng để giảm số lượng biến trong dữ liệu. Ví dụ: PCA, t-SNE.

## 1.2 Chuẩn bị dữ liệu:

Chia dữ liệu: Chia dữ liệu thành tập huấn luyện (training set) và tập kiểm tra (testing set) để đánh giá hiệu quả của mô hình.

Xử lý dữ liệu thiếu: Xử lý các giá trị bị thiếu bằng cách loại bỏ hoặc thay thế chúng.

Chuẩn hóa dữ liệu: Biến đổi dữ liệu để đảm bảo các biến có cùng thang đo hoặc phân phối.

## 1.3 Huấn luyện mô hình (Training the model):

Sử dụng tập dữ liệu huấn luyện để huấn luyện mô hình. Quá trình này liên quan đến việc tối ưu hóa các tham số của mô hình để giảm thiểu sai số dự đoán.

## 1.4 Đánh giá mô hình (Model Evaluation):

Đánh giá hiệu suất: Sử dụng các chỉ số như độ chính xác (accuracy), độ nhạy (recall), độ đặc hiệu (precision), F1 score, và MSE (Mean Squared Error) để đánh giá hiệu suất của mô hình trên tập dữ liệu kiểm tra.

Kỹ thuật kiểm định chéo (Cross-validation): Sử dụng để đánh giá mô hình một cách toàn diện hơn bằng cách chia dữ liệu thành nhiều tập con và kiểm tra mô hình trên mỗi tập con.

## 1.5 Tinh chỉnh và tối ưu hóa mô hình (Model Tuning and Optimization):

Tinh chỉnh tham số (Hyperparameter Tuning): Tối ưu hóa các tham số của mô hình để đạt được hiệu suất tốt nhất. Có thể sử dụng các phương pháp như Grid Search hoặc Random Search.

Chọn mô hình cuối cùng (Model Selection): Chọn mô hình có hiệu suất tốt nhất dựa trên các đánh giá và kiểm tra.

## 1.6 Triển khai mô hình (Model Deployment):

Sau khi chọn được mô hình tốt nhất, tiến hành triển khai mô hình vào hệ thống thực tế để sử dụng cho dự đoán và phân tích dữ liệu.

## 1.7 Theo dõi và bảo trì mô hình (Model Monitoring and Maintenance):

Theo dõi hiệu suất của mô hình sau khi triển khai và thực hiện các cập nhật cần thiết để đảm bảo mô hình vẫn hoạt động hiệu quả với dữ liệu mới.

# 2. Chia dữ liệu

## 2.1 Tập huấn luyện (Training Set)

Định nghĩa: Tập huấn luyện là phần dữ liệu được sử dụng để huấn luyện mô hình.

Mục đích: Mô hình học từ tập huấn luyện để tìm ra các mẫu hình và mối quan hệ trong dữ liệu.

Chi tiết: Tập huấn luyện thường chiếm phần lớn dữ liệu ban đầu, thường từ 60% đến 80% của toàn bộ dữ liệu.

## 2.2 Tập kiểm tra (Test Set)

Định nghĩa: Tập kiểm tra là phần dữ liệu được giữ lại để đánh giá hiệu suất của mô hình sau khi đã huấn luyện.

Mục đích: Đánh giá mô hình để đảm bảo rằng nó không chỉ hoạt động tốt trên tập huấn luyện mà còn trên dữ liệu mới, chưa từng thấy trước đó.

Chi tiết: Tập kiểm tra thường chiếm từ 20% đến 40% của toàn bộ dữ liệu ban đầu.

## 2.3 Tập xác nhận (Validation Set)

Định nghĩa: Tập xác nhận là một phần dữ liệu được tách ra từ tập huấn luyện để điều chỉnh các tham số của mô hình và lựa chọn mô hình tốt nhất.

Mục đích: Đánh giá mô hình trong quá trình huấn luyện và tinh chỉnh các tham số mà không ảnh hưởng đến tập kiểm tra.

Chi tiết: Tập xác nhận thường chiếm từ 10% đến 20% của tập huấn luyện.

## 2.4 Quy trình chia dữ liệu

Chia dữ liệu ban đầu: Tách dữ liệu ban đầu thành tập huấn luyện và tập kiểm tra.

Chia tập huấn luyện: Tách một phần của tập huấn luyện thành tập xác nhận (nếu cần).

## 2.5 Kỹ thuật kiểm định chéo (Cross-Validation)

Định nghĩa: Kỹ thuật kiểm định chéo là một phương pháp để đánh giá mô hình một cách toàn diện hơn bằng cách chia dữ liệu thành nhiều phần và sử dụng mỗi phần một lần làm tập kiểm tra và các phần còn lại làm tập huấn luyện.

Ví dụ phổ biến: K-Fold Cross-Validation là phương pháp chia dữ liệu thành K phần (folds) và lặp lại K lần, mỗi lần sử dụng một fold khác nhau làm tập kiểm tra và các fold còn lại làm tập huấn luyện.

## 2.6 Tại sao cần chia dữ liệu?

Tránh overfitting: Nếu chỉ sử dụng tập huấn luyện, mô hình có thể học quá kỹ các chi tiết và nhiễu trong dữ liệu huấn luyện, dẫn đến hiệu suất kém trên dữ liệu mới.

Đánh giá khách quan: Sử dụng tập kiểm tra giúp đánh giá mô hình một cách khách quan, vì nó không tham gia vào quá trình huấn luyện và tối ưu hóa.

Tinh chỉnh mô hình: Tập xác nhận giúp lựa chọn mô hình và tham số tốt nhất mà không cần sử dụng tập kiểm tra, giúp đảm bảo rằng tập kiểm tra chỉ được sử dụng cho đánh giá cuối cùng.

# 3. Demo 

## 3.1 Dùng thư viện scikit-learn

In [41]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score
from sklearn.preprocessing import StandardScaler
import joblib

Giả sử đã khám phá, tiền xử lý dữ liệu,... xong trước đó.

Bây giờ mình sẽ xây dựng mô hình để dự đoán nhiệt độ trung bình trong ngày ở TP HCM.

In [42]:
df = pd.read_csv('../data/HoChiMinh.csv')
df.head()

Address    Datetime  DatetimeEpoch  Tempmax  Tempmin  Temp   Dew  \
0  Hồ Chí Minh  2019-01-01     1546275600     28.0     24.0  26.1  18.9   
1  Hồ Chí Minh  2019-01-02     1546362000     25.4     24.0  24.7  21.0   
2  Hồ Chí Minh  2019-01-03     1546448400     30.6     23.4  27.1  21.7   
3  Hồ Chí Minh  2019-01-04     1546534800     34.0     25.5  29.4  21.8   
4  Hồ Chí Minh  2019-01-05     1546621200     33.6     25.0  29.1  21.0   

   Humidity  Precip  Precipprob  ...  Windgust Windspeed  Winddir  Pressure  \
0      64.9     3.1       100.0  ...      35.6      21.1      2.7    1012.2   
1      79.7    16.1       100.0  ...      47.2      16.2     11.1    1011.7   
2      72.9     0.1       100.0  ...      59.8      23.1     64.1    1011.9   
3      65.3     0.2       100.0  ...      56.2      29.5     95.3    1012.5   
4      63.4     0.0         0.0  ...      30.6      25.5    109.2    1012.3   

   Cloudcover  Visibility  Solarradiation  Solarenergy  Uvindex  Moonphase  
0        59.8        10.3            72.0          6.3      4.0       0.83  
1        61.5         9.4            52.4          4.5      3.0       0.87  
2        58.6        10.1           154.4         13.3      6.0       0.90  
3        43.4        10.3           209.4         18.2      7.0       0.94  
4        32.6        10.3           233.9         20.2      8.0       0.97  

[5 rows x 22 columns]

In [43]:
# bỏ các thuộc tính không cần khi train
df = df.drop(columns=['Address', 'Datetime', 'DatetimeEpoch', 'Preciptype'])
df.columns

Index(['Tempmax', 'Tempmin', 'Temp', 'Dew', 'Humidity', 'Precip', 'Precipprob',
       'Precipcover', 'Windgust', 'Windspeed', 'Winddir', 'Pressure',
       'Cloudcover', 'Visibility', 'Solarradiation', 'Solarenergy', 'Uvindex',
       'Moonphase'],
      dtype='object')

In [44]:
# Thuộc tính mình cần dự đoán
target = 'Temp'
# Lấy tên các thuộc tính
features = df.drop(columns=target).columns

X = df[features]
Y = df[target]

# 20% dữ liệu làm Test set
# seed random = 22120182
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=22120182)

# Chuẩn hóa dữ liệu
scaler = StandardScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)

# Xây dựng mô hình hồi quy tuyến tính
model = LinearRegression()
model.fit(X_train_normalized, Y_train)

# Dự đoán trên tập kiểm tra
Y_pred = model.predict(X_test_normalized)

In [45]:
# Đánh giá mô hình
mse = mean_squared_error(Y_test, Y_pred)
r2 = r2_score(Y_test, Y_pred)

print(f'Mean Squared Error: {mse:.5f}')
print(f'R^2 Score: {r2:.5f}')
print(Y_pred[:5])

Mean Squared Error: 0.02766
R^2 Score: 0.98451
[26.96005125 28.98920328 29.34711826 26.97029221 26.25300904]


## 3.2 Không dùng scikit-learn

Tham khảo: https://machinelearningcoban.com/2016/12/28/linearregression/

In [46]:
class LR:
    def __init__(self):
        self.w = None

    def fit(self, X, Y):
        X_b = np.c_[np.ones((X.shape[0], 1)), X]
        self.w = np.linalg.pinv(X_b.T.dot(X_b)).dot(X_b.T).dot(Y)

    def predict(self, X):
        X_b = np.c_[np.ones((X.shape[0], 1)), X]
        return X_b.dot(self.w)

    def mean_squared_error(self, Y_test, Y_pred):
        return ((Y_test - Y_pred) ** 2).mean()

    def r2_score(self, Y_test, Y_pred):
        return 1 - ((Y_test - Y_pred) ** 2).sum() / ((Y_test - Y_test.mean()) ** 2).sum()

Để cho tiện trong việc so sánh thì mình dùng lại $X$, $Y$ ở trên luôn :p, còn khi code mọi người có thể tự viết hàm chia training set, testing set với hàm chuẩn hóa.

In [47]:
model = LR()
model.fit(X_train_normalized, Y_train)
Y_pred = model.predict(X_test_normalized)

In [48]:
mse = model.mean_squared_error(Y_test, Y_pred)
r2 = model.r2_score(Y_test, Y_pred)

print(f'Mean Squared Error: {mse:.5f}')
print(f'R^2 Score: {r2:.5f}')
print(Y_pred[:5])

Mean Squared Error: 0.02766
R^2 Score: 0.98451
[26.96005125 28.98920328 29.34711826 26.97029221 26.25300904]


## 3.3 Đánh giá mô hình hồi quy (Evaluation)

**[MSE (Mean Squared Error)](https://scikit-learn.org/dev/modules/generated/sklearn.metrics.mean_squared_error.html)**: đo lường trung bình của bình phương sai số giữa giá trị thực và giá trị dự đoán. MSE là một thước đo để xác định mức độ dự đoán của mô hình sai lệch so với giá trị thực tế.

\begin{equation*}
MSE = \frac{1}{n}\sum_{i=1}^{n}(y_i-\hat{y}_i)^2
\end{equation*}

**[R^2](https://scikit-learn.org/dev/modules/generated/sklearn.metrics.r2_score.html)**: là một thước đo thống kê phản ánh tỷ lệ phương sai của biến đích được giải thích bởi các biến dự đoán trong mô hình hồi quy. Nó cung cấp một cách để xác định mức độ phù hợp của mô hình.

\begin{equation*}
R^2 = 1 - \frac{\sum_{i=1}^{n}(y_i - \hat{y}_i)^2}{\sum_{i=1}^{n}(y_i - \bar{y})^2}
\end{equation*}

Trong đó: 

- $n$: số mẫu.

- $y_i$: giá trị thực tế.

- $\hat{y}_i$: giá trị dự đoán.

- $\bar{y}$: giá trị trung bình. 

So sánh:

<table>
  <tr>
    <th style="text-align: center;">MSE</th>
    <th style="text-align: center;">R²</th>
  </tr>
  <tr>
    <td>- Đo lường lỗi dự đoán trung bình bình phương.</td>
    <td>- Đo lường tỷ lệ phương sai được giải thích bởi mô hình.</td>
  </tr>
  <tr>
    <td>- Giá trị càng nhỏ thì mô hình càng tốt.</td>
    <td>- Giá trị nằm trong khoảng từ 0 đến 1 (hoặc có thể âm nếu mô hình rất kém).</td>
  </tr>
  <tr>
    <td>- Khó so sánh trực tiếp giữa các mô hình khác nhau.</td>
    <td>- Dễ hiểu và so sánh giữa các mô hình khác nhau.</td>
  </tr>
</table>

---

So sánh kết quả khi dùng thư viện và không dùng thư viện thì thấy đều cho ra đáp án như nhau. Như vậy có thể coi là mình đã cài đặt chính xác phần modeling.

Đánh giá:

- MSE = 0.02766: Giá trị nhỏ này cho thấy sai số dự đoán trung bình của mô hình là khá thấp.

- R^2 = 0.98451: Giá trị này gần bằng 1, cho thấy mô hình có thể giải thích hầu hết phương sai trong dữ liệu mục tiêu.

Vậy với các giá trị MSE và R^2 như trên, có thể kết luận rằng mô hình hoạt động tốt trong việc dự đoán nhiệt độ trung bình.

\****Ở bước này thì mọi người có thể dùng nhiều model khác nhau rồi so sánh chúng để chọn ra model tốt nhất.***

---

In [49]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

# Giả sử bạn đã có dữ liệu X (đặc trưng) và y (nhiệt độ trung bình)
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=12345)

scaler = StandardScaler()
X_train_normalized = scaler.fit_transform(X_train)
X_test_normalized = scaler.transform(X_test)

# Khởi tạo mô hình Random Forest Regressor
model = RandomForestRegressor(n_estimators=100, random_state=54321)
model.fit(X_train, Y_train)
Y_pred = model.predict(X_test)

# Đánh giá
mse = mean_squared_error(Y_test, Y_pred)
r2 = r2_score(Y_test, Y_pred)
print(f'Mean Squared Error: {mse:.5f}')
print(f'R^2 Score: {r2:.5f}')
print(Y_pred[:5])
# print(list(Y_test[:5]))

Mean Squared Error: 0.14666
R^2 Score: 0.91020
[26.758 29.123 26.348 30.339 28.493]
